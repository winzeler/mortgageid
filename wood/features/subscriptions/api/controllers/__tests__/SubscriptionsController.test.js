const { Stripe } = require('stripe');
const moment = require('moment');
const request = require('supertest');
const MockDate = require('mockdate');
const nodemailerMock = require('nodemailer-mock');
const {
  connection,
  cleanDb,
  addDefaultUser,
  COOKIE_JWT_OWNER_ID_1,
  CSRF_TOKEN_USER_ID_1,
} = require('@wood/testHelper');
const fixtureStripeCoupons = require('@wood/config/stripe/__fixtures__/coupons.json.fixture');
const fixtureStripeProducts = require('@wood/config/stripe/__fixtures__/products.json.fixture');
const fixtureStripeTaxes = require('@wood/config/stripe/__fixtures__/taxes.json.fixture');
const { overrideConfig, overrideConfigFile, clearConfigOverrides } = require('#lib/Config');
const { AppBuilder } = require('#api/AppBuilder');

let app;
let db;
let server;
let agent;

describe('SubscriptionsController', () => {
  beforeAll(async () => {
    // Make sure correct feature controllers are initialized
    overrideConfig('app', 'features.app', []);
    overrideConfig('app', 'features.wood', ['users', 'subscriptions']);

    const builder = new AppBuilder({ mailer: nodemailerMock.createTransport({}) });
    app = await builder.getApp();
    db = app.get('db');
    server = await app.listen(4000);
    agent = request.agent(server);
  });

  beforeEach(async () => {
    MockDate.set('1/1/2020');
    await cleanDb();
    nodemailerMock.mock.reset();
    await addDefaultUser(db);
    overrideConfigFile('stripe/coupons', fixtureStripeCoupons);
    overrideConfigFile('stripe/products', fixtureStripeProducts);
    overrideConfigFile('stripe/taxes', fixtureStripeTaxes);
    overrideConfig('app', 'features', { app: [], wood: ['subscriptions', 'user'] });
  });

  afterEach(async () => {
    MockDate.reset();
    clearConfigOverrides();
  });

  afterAll(async () => {
    await connection.destroy();
    await server.close();
    await db.instance.$pool.end();
  });

  // ===============================================================================================

  describe('POST /subscriptions', () => {
    it('should validate empty request', async () => {
      const response = await agent
        .post('/api/subscriptions')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          line1: '',
          country: '',
          state: '',
          city: '',
          postal_code: '',
          payment_method_id: '',
          product_id: '',
          price_id: '',
        })
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should validate invalid country', async () => {
      overrideConfig('geography', 'countries', { VA: 'valid' });

      const response = await agent
        .post('/api/subscriptions')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          line1: 'line1',
          country: 'not-a-country',
          state: 'state',
          city: 'city',
          postal_code: 'postal code',
          payment_method_id: 'payment-method-id',
          product_id: 'prod_simple',
          price_id: 'price_simple_monthly',
        })
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should validate invalid state', async () => {
      overrideConfig('geography', 'countries', { VA: 'valid' });
      overrideConfig('geography', 'states', { VA: ['valid-state'] });

      const response = await agent
        .post('/api/subscriptions')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          line1: 'line1',
          country: 'VA',
          state: 'not-a-state',
          city: 'city',
          postal_code: 'postal code',
          payment_method_id: 'payment-method-id',
          product_id: 'prod_simple',
          price_id: 'price_simple_monthly',
        })
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should validate invalid product & price', async () => {
      overrideConfig('geography', 'countries', { VA: 'valid' });
      overrideConfig('geography', 'states', { VA: ['valid-state'] });

      const response = await agent
        .post('/api/subscriptions')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          line1: 'line1',
          country: 'VA',
          state: 'valid-state',
          city: 'city',
          postal_code: 'postal code',
          payment_method_id: 'payment-method-id',
          product_id: 'not-valid-product',
          price_id: 'not-valid-price',
        })
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    it('should validate price from incorrect product', async () => {
      overrideConfig('geography', 'countries', { VA: 'valid' });
      overrideConfig('geography', 'states', { VA: ['valid-state'] });

      const response = await agent
        .post('/api/subscriptions')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          line1: 'line1',
          country: 'VA',
          state: 'valid-state',
          city: 'city',
          postal_code: 'postal code',
          payment_method_id: 'payment-method-id',
          product_id: 'prod_simple',
          price_id: 'price_advanced_monthly',
        })
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should create subscription', async () => {
      overrideConfig('geography', 'countries', { CA: 'Canada' });
      overrideConfig('geography', 'states', { CA: ['Ontario'] });
      Stripe.prototype.customers = {
        create: jest.fn(() => ({ id: 'new_customer_id' })),
        update: jest.fn(() => ({})),
      };
      Stripe.prototype.paymentMethods = {
        attach: jest.fn(() => ({})),
      };
      Stripe.prototype.subscriptions = {
        create: jest.fn(() => ({
          id: 'new_subscription_id',
          status: 'active',
          current_period_end: moment().add(30, 'days').unix(),
          latest_invoice: {
            payment_intent: { id: 'latest_invoice_payment_intent' },
          },
        })),
      };
      Stripe.prototype.invoices = {
        retrieveUpcoming: jest.fn(() => ({ total: 1234 })),
      };

      const response = await agent
        .post('/api/subscriptions')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          line1: '290 Bremner Blvd',
          country: 'CA',
          state: 'Ontario',
          city: 'Toronto',
          postal_code: 'M5V 3L9',
          payment_method_id: 'payment-method-id',
          product_id: 'prod_simple',
          price_id: 'price_simple_monthly',
          coupon_id: '50_percent',
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();

      const team = await db.teams.findOne({ id: 1 });
      expect(team.stripe_customer_id).toBe('new_customer_id');

      const subscription = await db.subscriptions.findOne({ team_id: 1 });
      expect(subscription).toMatchSnapshot({
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
      });
    });
  });

  // ===============================================================================================

  describe('POST /subscriptions/retry', () => {
    it('should validate request', async () => {
      const response = await agent
        .post('/api/subscriptions/retry')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          line1: '',
          country: '',
          state: '',
          city: '',
          postal_code: '',
          payment_method_id: '',
          product_id: '',
          price_id: '',
        })
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should retry subscription creation, cancelling old sub with different values', async () => {
      await db.subscriptions.insert([{
        subscription_id: 'old_subscription_id',
        product_id: 'prod_advanced',
        price_id: 'price_advanced_monthly',
        coupon_id: '50_percent',
        tax_ids: '["txr_quebec"]',
        next_billing_date: moment().format(),
        status: 'incomplete',
        next_invoice_total: 1234,
        team_id: 1,
      }]);
      overrideConfig('geography', 'countries', { CA: 'Canada' });
      overrideConfig('geography', 'states', { CA: ['Ontario'] });
      Stripe.prototype.customers = {
        create: jest.fn(() => ({ id: 'new_customer_id' })),
        update: jest.fn(() => ({})),
      };
      Stripe.prototype.paymentMethods = {
        attach: jest.fn(() => ({})),
      };
      const delFn = jest.fn(() => {});
      Stripe.prototype.subscriptions = {
        del: delFn,
        create: jest.fn(() => ({
          id: 'new_subscription_id',
          status: 'active',
          current_period_end: moment().add(30, 'days').unix(),
          latest_invoice: { total: 4321 },
        })),
        retrieve: jest.fn(() => ({
          current_period_end: moment().add(30, 'days').unix(),
          status: 'active',
        })),
      };
      Stripe.prototype.invoices = {
        retrieveUpcoming: jest.fn(() => ({ total: 1234 })),
      };

      const response = await agent
        .post('/api/subscriptions/retry')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          line1: '290 Bremner Blvd',
          country: 'CA',
          state: 'Ontario',
          city: 'Toronto',
          postal_code: 'M5V 3L9',
          payment_method_id: 'payment-method-id',
          product_id: 'prod_simple',
          price_id: 'price_simple_monthly',
          coupon_id: '5_off',
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();
      expect(delFn).toHaveBeenCalled();

      const team = await db.teams.findOne({ id: 1 });
      expect(team.stripe_customer_id).toBe('new_customer_id');

      const subscription = await db.subscriptions.findOne({ team_id: 1 });
      expect(subscription).toMatchSnapshot({
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
      });
    });
  });

  // ===============================================================================================

  describe('GET /subscriptions/coupons/:id', () => {
    it('should get a valid coupon', async () => {
      Stripe.prototype.coupons = {
        retrieve: jest.fn(() => ({ times_redeemed: 49 })),
      };

      const response = await agent
        .get('/api/subscriptions/coupons/50_percent')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should error when missing coupon', async () => {
      const response = await agent
        .get('/api/subscriptions/coupons/missing')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .expect(404);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should error when too many redemptions', async () => {
      Stripe.prototype.coupons = {
        retrieve: jest.fn(() => ({ times_redeemed: 50 })),
      };

      const response = await agent
        .get('/api/subscriptions/coupons/50_percent')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should error when expired', async () => {
      const response = await agent
        .get('/api/subscriptions/coupons/expired')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should error when incorrect currency', async () => {
      const response = await agent
        .get('/api/subscriptions/coupons/canadian')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .expect(404);

      expect(response.body).toMatchSnapshot();
    });
  });

  // ===============================================================================================

  describe('GET /subscriptions/invoices', () => {
    it('should get list of invoices', async () => {
      await db.teams.update(1, { stripe_customer_id: 'cus_Ho12rn7QbDbZlD' });

      Stripe.prototype.invoices = {
        list: jest.fn(() => ({
          has_more: false,
          data: [{
            id: 'in_sample',
            created: 1596512754,
            currency: 'cad',
            invoice_pdf: 'https://some_url.com',
            status: 'draft',
            total: 1000,
          }],
        })),
      };

      const response = await agent
        .get('/api/subscriptions/invoices')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should return an empty list when no customer ID', async () => {
      const response = await agent
        .get('/api/subscriptions/invoices')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });

  // ===============================================================================================

  describe('DELETE /subscriptions', () => {
    it('should delete user\'s subscription', async () => {
      await db.subscriptions.insert([{
        subscription_id: 'subscription_id',
        product_id: 'prod_advanced',
        price_id: 'price_advanced_monthly',
        coupon_id: '50_percent',
        tax_ids: '[]',
        next_billing_date: moment().format(),
        status: 'active',
        next_invoice_total: 1234,
        team_id: 1,
      }]);

      const delFn = jest.fn(() => {});
      Stripe.prototype.subscriptions = {
        del: delFn,
      };

      await agent
        .delete('/api/subscriptions')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          message: '',
        })
        .expect(204);

      expect(delFn).toHaveBeenCalled();
      const subscription = await db.subscriptions.findOne({ team_id: 1 });
      expect(subscription).toBeNull();

      expect(nodemailerMock.mock.getSentMail()[0].text).toMatchSnapshot();
    });
  });

  // ===============================================================================================

  describe('POST /subscriptions/preview', () => {
    it('should validate the request', async () => {
      const response = await agent
        .post('/api/subscriptions/preview')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should preview subscription change', async () => {
      await db.subscriptions.insert([{
        subscription_id: 'subscription_id',
        product_id: 'prod_advanced',
        price_id: 'price_advanced_monthly',
        coupon_id: '50_percent',
        tax_ids: '[]',
        next_billing_date: moment().format(),
        status: 'active',
        next_invoice_total: 1234,
        team_id: 1,
      }]);

      Stripe.prototype.subscriptions = {
        retrieve: jest.fn(() => ({
          items: { data: [{ id: 'item_id' }] },
        })),
      };
      Stripe.prototype.invoices = {
        retrieveUpcoming: jest.fn(() => ({
          currency: 'usd',
          lines: {
            data: [{
              description: 'Unused time on Bronze after 21 Aug 2020',
              amount: -482,
            }, {
              description: '1 × Bronze (at $50.00 / year)',
              amount: 5000,
            }],
          },
          total_tax_amounts: [{
            amount: -520,
            inclusive: false,
            tax_rate: 'txr_canada',
          }],
          total_discount_amounts: [{
            amount: 500,
          }],
          discount: { coupon: { id: '50_percent' } },
          total: -4517,
        })),
      };

      const response = await agent
        .post('/api/subscriptions/preview')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          product_id: 'prod_advanced',
          price_id: 'price_advanced_monthly',
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });

  // ===============================================================================================

  describe('PUT /subscriptions', () => {
    it('should validate the request', async () => {
      const response = await agent
        .put('/api/subscriptions')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should change the subscription', async () => {
      await db.subscriptions.insert([{
        subscription_id: 'subscription_id',
        product_id: 'prod_advanced',
        price_id: 'price_advanced_monthly',
        coupon_id: '50_percent',
        tax_ids: '[]',
        next_billing_date: moment().format(),
        status: 'active',
        next_invoice_total: 1234,
        team_id: 1,
      }]);

      const updateFn = jest.fn(() => ({
        current_period_end: moment().add(30, 'days').unix(),
        status: 'active',
      }));
      Stripe.prototype.subscriptions = {
        retrieve: jest.fn(() => ({ items: { data: [{ id: 'item_id' }] } })),
        update: updateFn,
      };
      Stripe.prototype.invoices = {
        retrieveUpcoming: jest.fn(() => ({ total: -4517 })),
      };

      const response = await agent
        .put('/api/subscriptions')
        .set('Cookie', COOKIE_JWT_OWNER_ID_1)
        .set('X-CSRF-TOKEN', CSRF_TOKEN_USER_ID_1)
        .send({
          product_id: 'prod_simple',
          price_id: 'price_simple_monthly',
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();

      expect(updateFn).toHaveBeenCalled();

      const subscription = await db.subscriptions.findOne({ team_id: 1 });
      expect(subscription).toMatchSnapshot({
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
      });
    });
  });
});
