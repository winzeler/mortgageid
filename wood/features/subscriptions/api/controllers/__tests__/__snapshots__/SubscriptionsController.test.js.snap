// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubscriptionsController DELETE /subscriptions should delete user's subscription 1`] = `
"Message: 

User: name (user@email.com)"
`;

exports[`SubscriptionsController GET /subscriptions/coupons/:id should error when expired 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_EXPIRED",
      "title": "The coupon has expired.",
    },
  ],
}
`;

exports[`SubscriptionsController GET /subscriptions/coupons/:id should error when incorrect currency 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_NOT_FOUND",
      "title": "The coupon could not be found.",
    },
  ],
}
`;

exports[`SubscriptionsController GET /subscriptions/coupons/:id should error when missing coupon 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_NOT_FOUND",
      "title": "The coupon could not be found.",
    },
  ],
}
`;

exports[`SubscriptionsController GET /subscriptions/coupons/:id should error when too many redemptions 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_EXPIRED",
      "title": "The coupon has expired.",
    },
  ],
}
`;

exports[`SubscriptionsController GET /subscriptions/coupons/:id should get a valid coupon 1`] = `
Object {
  "data": Object {
    "coupon": Object {
      "amount_off": null,
      "currency": null,
      "duration": "forever",
      "duration_in_months": null,
      "id": "50_percent",
      "max_redemptions": 50,
      "metadata": Object {},
      "name": "50 percent",
      "percent_off": 50,
      "redeem_by": null,
    },
  },
}
`;

exports[`SubscriptionsController GET /subscriptions/invoices should get list of invoices 1`] = `
Object {
  "data": Object {
    "invoices": Array [
      Object {
        "created": 1596512754,
        "currency": "cad",
        "id": "in_sample",
        "invoice_pdf": "https://some_url.com",
        "status": "draft",
        "total": 1000,
      },
    ],
  },
}
`;

exports[`SubscriptionsController GET /subscriptions/invoices should return an empty list when no customer ID 1`] = `
Object {
  "data": Object {
    "invoices": Array [],
  },
}
`;

exports[`SubscriptionsController POST /subscriptions should create subscription 1`] = `
Object {
  "data": Object {
    "payment_intent": Object {
      "id": "latest_invoice_payment_intent",
    },
    "subscription": Object {
      "coupon_id": "50_percent",
      "id": 1,
      "next_billing_date": "2020-01-31T00:00:00+00:00",
      "next_invoice_total": 1234,
      "price_id": "price_simple_monthly",
      "product_id": "prod_simple",
      "status": "active",
      "subscription_id": "new_subscription_id",
      "tax_ids": Array [
        "txr_ontario",
      ],
      "team_id": 1,
      "trial_ends_at": null,
    },
  },
}
`;

exports[`SubscriptionsController POST /subscriptions should create subscription 2`] = `
Object {
  "coupon_id": "50_percent",
  "created_at": Any<Date>,
  "id": 1,
  "next_billing_date": 2020-01-31T00:00:00.000Z,
  "next_invoice_total": 1234,
  "price_id": "price_simple_monthly",
  "product_id": "prod_simple",
  "status": "active",
  "subscription_id": "new_subscription_id",
  "tax_ids": Array [
    "txr_ontario",
  ],
  "team_id": 1,
  "trial_ends_at": null,
  "updated_at": Any<Date>,
}
`;

exports[`SubscriptionsController POST /subscriptions should validate empty request 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "line1",
      },
      "title": "You must enter an address.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "country",
      },
      "title": "You must enter a country.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "state",
      },
      "title": "You must enter a state/province.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "city",
      },
      "title": "You must enter a city.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "postal_code",
      },
      "title": "You must enter a zip/postal code.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "payment_method_id",
      },
      "title": "You must choose a payment method.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "product_id",
      },
      "title": "You must select a plan.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "price_id",
      },
      "title": "You must select a price.",
    },
  ],
}
`;

exports[`SubscriptionsController POST /subscriptions should validate invalid country 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_NOT_IN_LIST",
      "meta": Object {
        "list": Object {
          "VA": "valid",
        },
      },
      "source": Object {
        "parameter": "country",
      },
      "title": "You must select a valid country.",
    },
  ],
}
`;

exports[`SubscriptionsController POST /subscriptions should validate invalid product & price 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_NOT_IN_LIST",
      "meta": Object {
        "list": Array [
          "Advanced",
          "Simple",
          "Member limit",
        ],
      },
      "source": Object {
        "parameter": "product_id",
      },
      "title": "You must select a valid product.",
    },
    Object {
      "code": "ERROR_NOT_IN_LIST",
      "meta": Object {
        "list": Array [],
      },
      "source": Object {
        "parameter": "price_id",
      },
      "title": "You must select a valid price.",
    },
  ],
}
`;

exports[`SubscriptionsController POST /subscriptions should validate invalid state 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_NOT_IN_LIST",
      "meta": Object {
        "list": Array [
          "valid-state",
        ],
      },
      "source": Object {
        "parameter": "state",
      },
      "title": "You must select a valid state / province.",
    },
  ],
}
`;

exports[`SubscriptionsController POST /subscriptions should validate price from incorrect product 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_NOT_IN_LIST",
      "meta": Object {
        "list": Array [
          "Simple Monthly",
          "Simple Annual",
        ],
      },
      "source": Object {
        "parameter": "price_id",
      },
      "title": "You must select a valid price.",
    },
  ],
}
`;

exports[`SubscriptionsController POST /subscriptions/preview should preview subscription change 1`] = `
Object {
  "data": Object {
    "discounts": Array [
      Object {
        "amount": "-$5.00 USD",
        "description": "50 percent (50%)",
      },
    ],
    "lines": Array [
      Object {
        "amount": "-$4.82 USD",
        "description": "Unused time on Bronze after 21 Aug 2020",
      },
      Object {
        "amount": "$50.00 USD",
        "description": "1 Ã— Bronze (at $50.00 / year)",
      },
    ],
    "taxes": Array [
      Object {
        "amount": "-$5.20 USD",
        "description": "GST (10%)",
      },
    ],
    "total": Object {
      "amount": "-$45.17 USD",
      "description": "Total credits to be applied",
    },
  },
}
`;

exports[`SubscriptionsController POST /subscriptions/preview should validate the request 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "product_id",
      },
      "title": "You must select a plan.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "price_id",
      },
      "title": "You must select a price.",
    },
  ],
}
`;

exports[`SubscriptionsController POST /subscriptions/retry should retry subscription creation, cancelling old sub with different values 1`] = `
Object {
  "data": Object {
    "subscription": Object {
      "coupon_id": "5_off",
      "id": 2,
      "next_billing_date": "2020-01-31T00:00:00+00:00",
      "next_invoice_total": 1234,
      "price_id": "price_simple_monthly",
      "product_id": "prod_simple",
      "status": "active",
      "subscription_id": "new_subscription_id",
      "tax_ids": Array [
        "txr_ontario",
      ],
      "team_id": 1,
      "trial_ends_at": null,
    },
  },
}
`;

exports[`SubscriptionsController POST /subscriptions/retry should retry subscription creation, cancelling old sub with different values 2`] = `
Object {
  "coupon_id": "5_off",
  "created_at": Any<Date>,
  "id": 2,
  "next_billing_date": 2020-01-31T00:00:00.000Z,
  "next_invoice_total": 1234,
  "price_id": "price_simple_monthly",
  "product_id": "prod_simple",
  "status": "active",
  "subscription_id": "new_subscription_id",
  "tax_ids": Array [
    "txr_ontario",
  ],
  "team_id": 1,
  "trial_ends_at": null,
  "updated_at": Any<Date>,
}
`;

exports[`SubscriptionsController POST /subscriptions/retry should validate request 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "line1",
      },
      "title": "You must enter an address.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "country",
      },
      "title": "You must enter a country.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "state",
      },
      "title": "You must enter a state/province.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "city",
      },
      "title": "You must enter a city.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "postal_code",
      },
      "title": "You must enter a zip/postal code.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "payment_method_id",
      },
      "title": "You must choose a payment method.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "product_id",
      },
      "title": "You must select a plan.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "price_id",
      },
      "title": "You must select a price.",
    },
  ],
}
`;

exports[`SubscriptionsController PUT /subscriptions should change the subscription 1`] = `
Object {
  "data": Object {
    "subscription": Object {
      "coupon_id": "50_percent",
      "id": 1,
      "next_billing_date": "2020-01-31T00:00:00+00:00",
      "next_invoice_total": -4517,
      "price_id": "price_simple_monthly",
      "product_id": "prod_simple",
      "status": "active",
      "subscription_id": "subscription_id",
      "tax_ids": Array [],
      "team_id": 1,
      "trial_ends_at": null,
    },
  },
}
`;

exports[`SubscriptionsController PUT /subscriptions should change the subscription 2`] = `
Object {
  "coupon_id": "50_percent",
  "created_at": Any<Date>,
  "id": 1,
  "next_billing_date": 2020-01-31T00:00:00.000Z,
  "next_invoice_total": -4517,
  "price_id": "price_simple_monthly",
  "product_id": "prod_simple",
  "status": "active",
  "subscription_id": "subscription_id",
  "tax_ids": Array [],
  "team_id": 1,
  "trial_ends_at": null,
  "updated_at": Any<Date>,
}
`;

exports[`SubscriptionsController PUT /subscriptions should validate the request 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "product_id",
      },
      "title": "You must select a plan.",
    },
    Object {
      "code": "ERROR_EMPTY",
      "source": Object {
        "parameter": "price_id",
      },
      "title": "You must select a price.",
    },
  ],
}
`;
